File: go-leetcode/dynamic-programming/416.分割等和子集.go
package leetcode

func canPartition(nums []int) bool {
	n := len(nums)
	if n < 2 {
		return false
	}

	// 计算总和
	sum := 0
	for _, v := range nums {
		sum += v
	}
	// 总和为奇数时不可能平分
	if sum%2 != 0 {
		return false
	}
	target := sum / 2

	// 如果最大元素超过 target，不可能
	maxNum := nums[0]
	for _, v := range nums {
		if v > maxNum {
			maxNum = v
		}
	}
	if maxNum > target {
		return false
	}

	// 创建 dp 数组，dp[i][j] 表示前 i 个元素能否凑出和为 j
	dp := make([][]bool, n)
	for i := 0; i < n; i++ {
		dp[i] = make([]bool, target+1)
		dp[i][0] = true // 和为 0 总是成立
	}

	// 初始化第一行
	if nums[0] <= target {
		dp[0][nums[0]] = true
	}

	// 动态规划
	for i := 1; i < n; i++ {
		for j := 1; j <= target; j++ {
			if nums[i] > j {
				dp[i][j] = dp[i-1][j]
			} else {
				dp[i][j] = dp[i-1][j] || dp[i-1][j-nums[i]]
			}
		}
	}
	return dp[n-1][target]
}

File: python-leetcode/dynamic-programming/416.分割等和子集.py
from typing import List


class Solution:
    def canPartition(self, nums: List[int]) -> bool:
        n = len(nums)
        if n < 2:
            return False

        total = sum(nums)
        # 总和为奇数无法平分
        if total % 2 != 0:
            return False
        target = total // 2

        # 最大元素超过目标值则无法平分
        if max(nums) > target:
            return False

        # 创建 dp 表，dp[i][j] 表示前 i 个元素能否凑出和为 j
        dp = [[False] * (target + 1) for _ in range(n)]
        for i in range(n):
            dp[i][0] = True  # 和为 0 总是可以做到

        # 初始化第一行
        if nums[0] <= target:
            dp[0][nums[0]] = True

        # 填充 dp 表
        for i in range(1, n):
            for j in range(1, target + 1):
                if nums[i] > j:
                    dp[i][j] = dp[i - 1][j]
                else:
                    dp[i][j] = dp[i - 1][j] or dp[i - 1][j - nums[i]]

        return dp[n - 1][target]
